### PR TITLE
repo,baseplugin: support trusting repo keys (#2437)

### PR DESCRIPTION
**Pull request by artivis**
Backport #2437 to `legacy` branch.
LP: #1811304

Original commit:

Today, snapcraft allows insecure repositories (i.e. repos without a key)
because that is the default behavior in Xenial (core16). This behavior
changed in Zesty, and stays changed in Bionic (core18). Add the ability
to trust keys to the plugin API, take advantage of it in the Catkin
plugin, and flip the switch to only allow snapcraft to use secure
repositories.

LP: #1811304

Signed-off-by: Kyle Fazzari <kyrofa@ubuntu.com>
Signed-off-by: artivis <jeremie.deray@canonical.com>

- [ ] Have you followed the [guidelines for contributing](https://github.com/snapcore/snapcraft/blob/master/CONTRIBUTING.md)?
- [ ] Have you signed the [CLA](http://www.ubuntu.com/legal/contributors/)?
- [x] Have you successfully run `./runtests.sh static`?
       Couldn't get to install `black`
- [x] Have you successfully run `./runtests.sh tests/unit`?
       `xdelta3`-dependent tests are failing, (re)installing `xdelta3` wouldn't fix it.

-----

### CODE REVIEW

**Comment by kyrofa** [comment]
Tests are failing here because this background is still using the old key:

```
$ gpg --no-default-keyring --keyring ./keyrings/ros.gpg --list-keys
./keyrings/ros.gpg
------------------
pub   dsa1024 2009-12-24 [SC]
      421C365BD9FF1F717815A3895523BAEEB01FA116
uid           [ unknown] ROS Builder <rosbuild@ros.org>
sub   elg2048 2009-12-24 [E]
```

Grab the one from master for this PR. It should look like this:

```
$ gpg --no-default-keyring --keyring ./keyrings/ros.gpg --list-keys
./keyrings/ros.gpg
------------------
pub   rsa4096 2019-05-30 [SC] [expires: 2021-05-29]
      C1CF6E31E6BADE8868B172B4F42ED6FBAB17C654
uid           [ unknown] Open Robotics <info@osrfoundation.org>
```

**Comment by artivis** [comment]
Keyring updated from master :+1: .

**Comment by sergiusens** [comment]
Great, seems something needs to be done about ament too (which is not part of the backport as ament is not available on Snapcraft 3), should this be a new PR or additional commit?

**Comment by artivis** [comment]
> Great, seems something needs to be done about ament too (which is not part of the backport as ament is not available on Snapcraft 3), should this be a new PR or additional commit?

This likely should be part of this PR, I'll get to it in any case and wait for @kyrofa's opinion.

**Comment by artivis** [comment]
After a look on the `amend` side, it is not a quick fix. I'll wait for further notice before attempting to fix it.

**Comment by kyrofa** [comment]
> something needs to be done about ament too

Good catch, I forgot about that one. Ideally it would be a new PR, but they are dependent on each other in order for the tests to pass, so might as well do it (as another commit, please) here.

**Comment by artivis** [comment]
> so might as well do it (as another commit, please) here.

Ok I'm on it.

**Comment by artivis** [comment]
Any chance I can run the failing test on my machine ?

**Comment by sergiusens** [comment]
To run that specific test, having built the snap of snapcraft with your changes. Obtain spread and from an environment that has LXD setup just run:

    spread -debug lxd:ubuntu-16.04/tests/spread/plugins/ament/run

from the project root

File: snapcraft/plugins/ament.py [hunk 1]
```diff
@@ -105,6 +105,7 @@ def __init__(self, name, options, project):
             version=self.options.version,
             bootstrap_path=self._bootstrap_dir,
             ubuntu_sources=self.PLUGIN_STAGE_SOURCES,
+            ubuntu_keyrings=self.PLUGIN_STAGE_SOURCES,
```
Comment (kyrofa) [inline snapcraft/plugins/ament.py]: This isn't quite the [commit I've verified](https://github.com/artivis/snapcraft/commit/c1edb60e69540e7f380b3a757e6439ea5695d0f8). It needs the keyring.

Comment (artivis) [reply snapcraft/plugins/ament.py]: My bad I'm messed-up pushing an outdated commit.

**Comment by artivis** [comment]
CI job is failing on `The job exceeded the maximum time limit for jobs, and has been terminated`.

File: tests/spread/plugins/catkin/snaps/ros-talker-listener/snap/snapcraft.yaml [hunk 2]
```diff
@@ -3,6 +3,7 @@ version: '1.0'
 summary: ROS Example
 description: Contains talker/listener ROS packages and a .launch file.
 confinement: strict
+base: core18
```
Comment (kyrofa) [inline tests/spread/plugins/catkin/snaps/ros-talker-listener/snap/snapcraft.yaml]: Was this part of the original backport? I don't believe it needs to be here, legacy won't do anything meaningful with it.

Comment (artivis) [reply tests/spread/plugins/catkin/snaps/ros-talker-listener/snap/snapcraft.yaml]: This is [on my other branch too](https://github.com/artivis/snapcraft/commit/c76ec02dd2957d21bb7490f2a760753637e0a12d#diff-c1b90dabf4e9f14a0aa8d12648020edbR6) and was copied from [here](https://github.com/snapcore/snapcraft/pull/2437/files#diff-c1b90dabf4e9f14a0aa8d12648020edbR6). So i'd imagine it was initially there. In any case I'm removing it :+1:.

Comment (kyrofa) [reply tests/spread/plugins/catkin/snaps/ros-talker-listener/snap/snapcraft.yaml]: Ah, yep, makes sense. Yeah, toast it here.

Comment (artivis) [reply tests/spread/plugins/catkin/snaps/ros-talker-listener/snap/snapcraft.yaml]: Done.

**Pull request merged**